src/pytorch_kinematics/chain.py

```diff
@@ -177,40 +449,39 @@ def _generate_serial_chain_recurse(root_frame, end_frame_name):
-
-    def forward_kinematics(self, th, world=None, end_only=True):
-        if world is None:
-            world = tf.Transform3d()
-        if world.dtype != self.dtype or world.device != self.device:
-            world = world.to(dtype=self.dtype, device=self.device, copy=True)
-        th, N = ensure_2d_tensor(th, self.dtype, self.device)
-        zeros = torch.zeros([N, 1], dtype=world.dtype, device=world.device)
-
-        theta_idx = 0
-        link_transforms = {}
-        trans = tf.Transform3d(matrix=world.get_matrix().repeat(N, 1, 1))
-        for f in self._serial_frames:
-            if f.link.offset is not None:
-                trans = trans.compose(f.link.offset)
-
-            if f.joint.joint_type == "fixed":  # If fixed
-                trans = trans.compose(f.get_transform(zeros))
-            else:
-                trans = trans.compose(f.get_transform(th[:, theta_idx].view(N, 1)))
-                theta_idx += 1
-
-            link_transforms[f.link.name] = trans
-
-        return link_transforms[self._serial_frames[-1].link.name] if end_only else link_transforms
-
     def jacobian(self, th, locations=None):
         if locations is not None:
             locations = tf.Transform3d(pos=locations)
         return jacobian.calc_jacobian(self, th, tool=locations)
+
+    def forward_kinematics(self, th, end_only: bool = True):
+        """ Like the base class, except `th` only needs to contain the joints in the SerialChain, not all joints. """
+        frame_indices, th = self.convert_serial_inputs_to_chain_inputs(end_only, th)
+
+        mat = super().forward_kinematics(th, frame_indices)
+
+        if end_only:
+            return mat[self._serial_frames[-1].name]
+        else:
+            return mat
+
+    def convert_serial_inputs_to_chain_inputs(self, end_only, th):
+        if end_only:
+            frame_indices = self.get_frame_indices(self._serial_frames[-1].name)
+        else:
+            # pass through default behavior for frame indices being None, which is currently
+            # to return all frames.
+            frame_indices = None
+        if get_th_size(th) < self.n_joints:
+            # if th is only a partial list of joints, assume it's a list of joints for only the serial chain.
+            partial_th = th
+            nonfixed_serial_frames = list(filter(lambda f: f.joint.joint_type != 'fixed', self._serial_frames))
+            if len(nonfixed_serial_frames) != len(partial_th):
+                raise ValueError(f'Expected {len(nonfixed_serial_frames)} joint values, got {len(partial_th)}.')
+            th = torch.zeros(self.n_joints, device=self.device, dtype=self.dtype)
+            for frame, partial_th_i in zip(nonfixed_serial_frames, partial_th):
+                k = self.frame_to_idx[frame.name]
+                jnt_idx = self.joint_indices[k]
+                if frame.joint.joint_type != 'fixed':
+                    th[jnt_idx] = partial_th_i
+        return frame_indices, th
```

Comment:
This function doesn't seem to handle batch inputs - SerialChain.forward_kinematics is not currently working for me for batched inputs.

Reply:
That's odd, because we have many tests with batch inputs. What's the shape of `th` you're using? and which model?

Reply:
I am using a model of victor, so the full chain has 14 DoF, but doing FK with on a single arm, so I am using an input `th` of `N x 7`. It looks like the logic for when there is only a partial list of joints fails for batched inputs. I will put a fix in a new suggestion. Probably worth including a test for this case.

Reply:
Can you re-test this? I added a new test but I just want to check it works for your use case